Add snmp_trap input plugin

reimda opened the pull request:
### Required for all PRs:

- [ x ] Signed [CLA](https://influxdata.com/community/cla/).
- [ x ] Associated README.md updated.
- [ x ] Has appropriate unit tests.

closes #4377

reimda commented:
to do:
- Mock OID name resolution so unit tests pass even when snmptranslate isn't available
- Add v1 traps.  Version 2c traps seem to work now.
- Format values better, possibly using the same mechanism as the snmp plugin or by adding support for textual convention in the MIB.
- Check on upstream support for inform.

glinton commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+		//no ip means listen on all interfaces, ipv4 and ipv6
+		err := s.listener.Listen(":" + strconv.FormatUint(uint64(s.Port), 10))
+		if err != nil {
+			s.Errch <- err
```
This channel isn't read anywhere, nor initialized, probably want to address.

reimda replied:
It's meant to be used in the test but it's clearly not working like I wanted if it's not initialized :)

glinton commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+		err := s.listener.Listen(":" + strconv.FormatUint(uint64(s.Port), 10))
+		if err != nil {
+			s.Errch <- err
+			log.Panicf("error in listen: %s", err)
```
I don't believe `log.Panic` is a preferred method of logging in telegraf

reimda replied:
Switching to log.Errorf(), thanks

glinton commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+	s.wg.Wait()
+}
+
+func makeTrapHandler(s *SnmpTrap) func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr) {
```
This feels odd. Was there a reason not to simply pass in the timefunc and accumulator?

reimda replied:
Passing in timefunc and accumulator would work for the code as it is.  I expect to need more info from the SnmpTrap struct in the future, like settings on how to interpret or convert data from the trap.  I think it's going to be enough variables that it will make sense to have the whole SnmpTrap struct available.

Would it be less odd to have makeTrapHandler be a method of SnmpTrap?

glinton commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+			value := v.Value
+
+			//use system mibs to resolve the name if possible
+			_, _, oidText, _, err := snmp.SnmpTranslate(v.Name)
```
Did you mean to use `name` here and on line 128 instead of v.Name?

reimda replied:
I meant to use v.Name in both places but the logic here could be clearer.  I think I'll rearrange things a little.

glinton commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+		for _, v := range packet.Variables {
+			//build a name and value for each variable to use as tags
+			//and fields.  defaults are the uninterpreted values
+			name := v.Name
```
I'm not sure what is returned as a name here, but would it be preferred to snake case them to fit better as field/tag keys?

reimda replied:
v.Name is a numeric OID, something like ".1.3.6.1.6.3.1.1.5.1".  If MIB lookup is set up we'll get something like "coldStart" that we use instead.  SNMP convention is camel case and well defined in many MIBs over many decades.  I was going to leave field names as camel case.  If there's enough demand we could consider a configuration option to automatically convert to snake case.

danielnelson replied:
Normally we use snake_case in Telegraf, even when the source system uses another style, but for this plugin I think we should mirror the snmp input behavior.

glinton commented on Gopkg.lock:
```diff
@@ -513,6 +513,14 @@
 
+[[projects]]
+  digest = "1:68c64bb61d55dcd17c82ca0b871ddddb5ae18b30cfe26f6bfd4b6df6287dc2e0"
+  name = "github.com/golang/mock"
```
Looks like this is unused, is this still in the plans to use?

reimda replied:
It's a new dependency of gosnmp.  I updated to latest gosnmp to get a recent change related to printing the type of a trap variable.

glinton commented on plugins/inputs/snmp_trap/snmp_trap_test.go:
```diff
@@ -0,0 +1,149 @@
+)
+
+func TestTranslate(t *testing.T) {
+	mibName, oidNum, oidText, conversion, err := snmp.SnmpTranslate(".1.3.6.1.6.3.1.1.5.1")
```
Is there a reason this test isn't in `inputs/snmp/snmp_test.go`, as it only tests a function from that package?

reimda replied:
I put it here so these tests fail quickly if somebody changes snmp.SnmpTranslate() without knowing that it's used by snmp_trap.

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap_test.go:
```diff
@@ -0,0 +1,149 @@
+
+// TestReceiveTrap
+func TestReceiveTrap(t *testing.T) {
+	const port = 12399 //todo: find unused port
```
Use port 0 to autoselect an unused port.

glinton replied:
Looks like he's trying to send to the that socket though, and it's currently not plumbed to get the port back that go selects.

reimda replied:
Listening on port 0 would be ideal but I couldn't see a way to do it without changing gosnmp.  Since gosnmp uses net.ListenUDP() listening on port 0 should work, but (*TrapListener) Listen() doesn't expose the connection you need to get which port was autoselected.

reimda commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+	s.listener.Params = gosnmp.Default
+	//s.listener.Params.Logger = log.New(os.Stdout, "", 0)
+
+	//wrap the handler, used in unit tests
```
Thanks.  It looks like I've been doing it wrong all this time :)

reimda commented on plugins/inputs/snmp/snmp.go:
```diff
@@ -948,7 +948,7 @@ var snmpTranslateCachesLock sync.Mutex
 
 // snmpTranslate resolves the given OID.
-func snmpTranslate(oid string) (mibName string, oidNum string, oidText string, conversion string, err error) {
+func SnmpTranslate(oid string) (mibName string, oidNum string, oidText string, conversion string, err error) {
```
It seems natural that snmp_trap would depend on snmp.  I don't see any benefit to moving the code to a common dir.

I generally don't like the idea of common dirs. It seems like they always hold stuff that would fit better elsewhere.

danielnelson replied:
Perhaps, we have some things that are either shared by many plugins, mostly these are in `internal` right now, which ought to be moved outside of this special directory so that they may be used by plugins outside of the repo.  I would be inclined to agree it doesn't make sense to have these in a common directory, but I don't want too many top level directories.  We also have some functions that are used from both, for example the kafka input and the kafka output where it's less clear which plugin should own the logic.

reimda commented on plugins/inputs/snmp_trap/README.md:
````diff
@@ -0,0 +1,33 @@
+### Configuration
+```toml
+  ## Port to listen on.  Default 162
+  #port = 162
````
good idea, I'll do this

reimda commented on plugins/inputs/snmp_trap/README.md:
````diff
@@ -0,0 +1,33 @@
+
+### Example Output
+```
+> snmp_trap,host=debian,trap_name=coldStart,trap_version=2c sysUpTimeInstance_type="TimeTicks",snmpTrapEnterprise.0="linux",snmpTrapEnterprise.0_type="ObjectIdentifier",sysUpTimeInstance=1i 1573078928344595213
````
ok, thanks

reimda commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+
+var sampleConfig = `
+  ## Port to listen on.  Default 162
+  #port = 162
```
This would require a change to gosnmp.  "udp" is hard coded in (t *TrapListener) Listen().

https://github.com/soniah/gosnmp/blob/master/trap.go#L154

reimda commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+	s.listener = gosnmp.NewTrapListener()
+	s.listener.OnNewTrap = makeTrapHandler(s)
+	s.listener.Params = gosnmp.Default
+	//s.listener.Params.Logger = log.New(os.Stdout, "", 0)
```
Yep, I'll clear them out before the real PR

reimda commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,144 @@
+		defer s.wg.Done()
+
+		//no ip means listen on all interfaces, ipv4 and ipv6
+		err := s.listener.Listen(":" + strconv.FormatUint(uint64(s.Port), 10))
```
I'll do this.  The test uses Listening() to make sure things are ready before sending a trap.  It will be nicer to move that code from the test into Start().  It will get rid of Errch too.

tesibelda commented:
Great job reimda. Some suggestions:
 - it would be useful to have a snmp_mib tag added. It would be helpfull for post processing filtering of these points (say with kapacitor). When using a large amount of MIBs, some of them might use the same string for snmp_trap.
  - it would be useful to have a snmp_host tag added. It would include the DNS resolved name of the IP address of device sending the trap. This would help identify the device having the issue reported by the trap.

reimda commented:
Thanks for the feedback @tesibelda 

>     * it would be useful to have a snmp_mib tag added ...

I added this as a tag called trap_mib to match the existing tag trap_name.

>     * it would be useful to have a snmp_host tag added. It would include the DNS resolved name of the IP address of device sending the trap ...

Since we already have the numeric IP address, I added it as a tag called source.  Other inputs use the name "source" for this so it makes sense to follow convention.  I agree a DNS name could be more useful than the IP but I'd like to wrap up this PR.  Adding a setting to do reverse DNS sounds like a great follow-on feature.

reimda commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,147 @@
+
+			// use system mibs to resolve the name if possible
+			_, _, oidText, _, err := snmp.SnmpTranslate(v.Name)
+			if nil == err {
```
Wouldn't dropping it entirely make it harder to set up and debug?  How about always adding the oid in one tag and optionally adding other tags for mib/name if they're available?  Then even if someone doesn't have snmptranslate and net-snmp set up right, at least won't have missing data.

danielnelson replied:
It would be a bit harder to debug, you would need to go look at the logfile of the Telegraf instance.  We do have a internal metric for errors, so we could increment this (currently done by logging an error or using acc.AddError).

I still think we should skip the metrics in spite of this, because there isn't any way to clean up bad data in the database.  Once it has been added, it is essentially permanent until and unless the retention policy removes it.

danielnelson replied:
> Once it has been added, it is essentially permanent until and unless the retention policy removes it.

I guess this isn't completely true, you could always use:
```
drop series from snmp_tap where 'oid' == '0.1.2.3'
```

reimda replied:
I'll change it to drop the trap and log a warning.  I agree putting data into the database with numeric

reimda commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,147 @@
+		// no ip means listen on all interfaces, ipv4 and ipv6
+		err := s.listener.Listen(s.ServiceAddress)
+		if err != nil {
+			s.Log.Errorf("error in listen: %s", err)
```
Would it be better to have Start() return the error and not log anything?

danielnelson replied:
Yeah

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,147 @@
+		}
+	}()
+
+	<-s.listener.Listening()
```
Add a info log message with the listening address and port, keeping style with **most** of the other listening plugins.  For example:
```
2019-11-15T00:04:11Z I! [inputs.socket_listener] Listening on udp://0.0.0.0:8094
```

danielnelson commented on plugins/inputs/snmp_trap/README.md:
```diff
@@ -0,0 +1,40 @@
+- snmp_trap
+  - tags:
+	- source (string, IP address of trap source)
+	- trap_name (string, value from SNMPv2-MIB::snmpTrapOID.0 PDU)
```
Update names here

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,251 @@
+}
+
+func realExecCmd(arg0 string, args ...string) ([]byte, error) {
+	return exec.Command(arg0, args...).Output()
```
Use the functions in `internal/exec.go` with a timeout.

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,251 @@
+		s.listener.OnNewTrap = s.makeHandlerWrapper(s.listener.OnNewTrap)
+	}
+
+	split := strings.SplitN(s.ServiceAddress, "://", 2)
```
I'd have probably used `url.Parse`

reimda replied:
Thanks for the suggestion.  I'll leave it because it's about as much work to use url.Parse and then build a new string with the host and port.  If we ever support other transports we should switch to url.Parse

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,251 @@
+
+func (s *SnmpTrap) Stop() {
+	s.listener.Close()
+	_ = <-s.errCh
```
Not required, will be garbage collected, this would allow you to make the channel a local variable.

reimda replied:
Instead of making s.errCh local to Start() I left it in the SnmpTrap struct and changed Stop() to log if there's an error.

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,251 @@
+	var out []byte
+	out, err = s.execCmd("snmptranslate", "-Td", "-Ob", "-m", "all", oid)
+
+	//var e mibEntry
```
delete, also line 239

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,251 @@
+		return e, err //Errorf(scanner.Err(), "getting OID text")
+	}
+
+	e.oidText = scanner.Text()
```
Optional, I wonder if it would be good to use fmt.Scanf for this text?

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,251 @@
+
+			name := e.oidText
+
+			fields[name] = value
```
What are the possible types?  Do we need to do any conversions?

reimda replied:
I suspect we'll need to do more work converting and formatting.  To do that I need some real world traps to test with.  I was hoping we could get this out and ask people watching the feature request for test cases.

danielnelson commented on plugins/inputs/snmp_trap/snmp_trap.go:
```diff
@@ -0,0 +1,251 @@
+				var e mibEntry
+				var err error
+				if !ok {
+					s.Log.Errorf("Error getting value OID: %v", err)
```
`err` is going to be nil here...
